File: src/CPP/src/leetcode/LargestRectangleInHistogram.cpp
/*
 * [Source] https://leetcode.com/problems/largest-rectangle-in-histogram/
 * [Difficulty]: Hard
 * [Tag]: Array
 * [Tag]: Stack
 */

#include <iostream>
#include <vector>
#include <stack>
#include <functional>
#include <climits>

using namespace std;

// [Solution]: Since the area depends min height on the path. Use a stack to track previous min height index.
// [Corner Case]: stack is empty. Dup height. heights are increasing
class Solution {
private:
    int findMaxArea(stack<size_t>& idxes, const vector<int>& heights, size_t ed, int nextHeight) {
        int maxArea = 0;
        while (!idxes.empty() && nextHeight < heights[idxes.top()]) {
            int height = heights[idxes.top()];
            idxes.pop();
            size_t st = idxes.empty() ? 0 : idxes.top() + 1;
            int width = ed - st;
            maxArea = max(maxArea, height * width);
        }
        return maxArea;
    }

public:
    int largestRectangleArea(vector<int>& heights) {
        if (heights.size() == 0)
            return 0;

        int maxArea = 0;
        stack<size_t> idxes;
        for (size_t i = 0; i < heights.size(); ++i) {
            if (idxes.size() == 0 || heights[i] > heights[idxes.top()]) {
                idxes.push(i);
                continue;
            }
            maxArea = max(maxArea, findMaxArea(idxes, heights, i, heights[i]));

            idxes.push(i);
        }
        maxArea = max(maxArea, findMaxArea(idxes, heights, heights.size(), 0));

        return maxArea;
    }
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/datastructure/LargestRectangleInHistogram.java
 */

/* Java solution
public class Solution {
    public int largestRectangleArea(int[] height) {
		int len=height.length;
		if(len==0) return 0;
		int max=0;
		Stack<Integer> heiStack=new Stack<Integer>();
		Stack<Integer> idxStack=new Stack<Integer>();
		for(int i=0; i<len; i++){
			if(heiStack.empty() || height[i]>=heiStack.peek()){
				heiStack.push(height[i]);
				idxStack.push(i);
			}else{ // Compute
				int l=0;
				while(!heiStack.empty() && heiStack.peek()>height[i]){ // Not equal to leave the equal situation to next compute
					l=idxStack.pop();
					max=Math.max(max, (i-l)*heiStack.pop()); // i already bigger than the current right bound 
				}
				heiStack.push(height[i]);
				idxStack.push(l); // This rectangle should computed before encounter the prev left bound, otherwise wrong when l=0
			}
		}
		while(!heiStack.empty()){
			max=Math.max(max, (len-idxStack.pop())*heiStack.pop());
		}
		return max;
    }
}
 */

int main() {
    Solution sol;

    vector<int> heights = {2,1,5,5,6, 2,3};
    cout << sol.largestRectangleArea(heights) << endl;

    return 0;
}

File: wwwroot/js/largest-rectangle-in-histogram.js
/**
 * @param {number[]} heights
 * @return {number}
 */
var largestRectangleAreaDP = function(heights) {
    // area = minHeight(x, y) * (y - x + 1)
    // minHeight(x, y) = min(minHeight(x, y - 1), height(y))
    let m = heights.length;
    if (m === 0) {
        return 0;
    }

    let matrix = new Array(m);
    for (let i = 0; i < m; i++) {
        matrix[i] = new Array(m);
    }

    let maxArea = 0;
    for (let i = 0; i < m; i++) {
        for (let j = i; j < m; j++) {
            if (j === i) {
                matrix[i][j] = heights[i];
            } else {
                matrix[i][j] = Math.min(matrix[i][j - 1], heights[j]);
            }

            maxArea = Math.max(maxArea, matrix[i][j] * (j - i + 1));
        }
    }

    return maxArea;
};

/**
 * @param {number[]} heights
 * @return {number}
 */
 var largestRectangleArea = function(heights) {
    // use stack to record previous min height
    let m = heights.length;
    if (m === 0) {
        return 0;
    }

    let hStack = [-1];
    let iStack = [-1];
    let maxArea = 0;
    let i = 0;
    while (i <= m) {
        if (i === m || hStack[hStack.length - 1] > heights[i]) {
            if (hStack.length === 1) {
                break;
            }

            let h = hStack.pop();
            iStack.pop();
            maxArea = Math.max(h * (i - 1 - iStack[iStack.length - 1]), maxArea);
        } else {
            // Equals need push, to deal with the wave case: [0,1,0,1]
            hStack.push(heights[i]);
            iStack.push(i);
            i++;
        }
    }

    return maxArea;
};

document.getElementById("largest-rectangle-in-histogram").innerHTML =
    // largestRectangleArea([2,1,5,5,6,2,3]);
    // largestRectangleArea([2,4]);
    // largestRectangleArea([2,1,2]);
    // largestRectangleArea([0,9]);
    largestRectangleArea([0,1,0,1]);
